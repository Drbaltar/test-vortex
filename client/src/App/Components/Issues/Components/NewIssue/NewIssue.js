/*
*  This module serves as a React Component that 
*  displays the view for submitting new issues
*  Author: Kyle McCain
*  Date: 10 December 2020
*/

import React from 'react';
import Axios from 'axios';

import IssueForm from '../IssueForm/IssueForm';
import SuccessMessage from '../../../shared-components/SuccessMessage/SuccessMessage';

class NewIssue extends React.Component {
    constructor(props) {
        super(props);

        this.initialState = {
            submissionResponse: '',
            successAlert: false
        };

        this.state = this.initialState;
    }

    submitNewIssue = (issueData) => {
        Axios.post('/api/issues/submit', issueData)
            .then((response) => this.setSuccesMessage(response))
            .catch((error) => {
                if(error.response) {
                    this.setErrorMessage(error.response.data.message);
                } else if (error.request) {
                    this.setErrorMessage( 
                        'There was an error submitting your issue! ' +
                        'Verify you have a valid Internet connection and try to resubmit.')
                } else {
                    this.setErrorMessage(  
                        'There was an internal client error when trying to submit your issue! ' +
                        'Please try to resubmit.')
                }
            })
    }

    setSuccesMessage = (message) => {
        this.setState({submissionResponse: message, successAlert: true})
    }

    setErrorMessage = (message) => {
        this.setState({submissionResponse: message})
    }

    getNewIssueView = () => {
        return (
            <IssueForm submitEvent={(issueData) => this.submitNewIssue(issueData)}
                submitButtonText='Submit' cancelButtonText='Clear All'
                errorMessage={this.state.submissionResponse}
                clearErrorMessage={() => this.clearErrorMessage()}/>
        )
    }

    getSuccessMessage = () => {
        return (
            <SuccessMessage message={this.state.submissionResponse.data}
                clickHandler={() => this.returnToForm()}/>
        );
    }

    clearErrorMessage = () => {
        this.setState({submissionResponse: ''})
    }

    returnToForm = () => {
        this.setState(this.initialState);
    }

    render() {
        return(
            <div>
                {this.state.successAlert ? this.getSuccessMessage() : this.getNewIssueView()}
            </div>
        );
    }
}

export default NewIssue;