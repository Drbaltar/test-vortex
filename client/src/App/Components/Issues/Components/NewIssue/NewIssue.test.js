import React from 'react';
import { shallow } from 'enzyme';
import NewIssue from '../NewIssue/NewIssue';

describe('NewIssue', () => {
    const wrapper = shallow(<NewIssue>

    </NewIssue>);

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('renders the SubmissionBody component', () => {
        expect(wrapper.find('SubmissionBody').exists()).toBe(true);
    });

    it('passes the IssueForm component to the SubmissionBody component', () => {
        expect(wrapper.find('SubmissionBody').children().name()).toEqual('IssueForm');
    });
    
    it('passes the title prop to the IssueForm component', () => {
        expect(wrapper.find('IssueForm').prop('title')).toEqual('Submit New Issue');
    });

    it('passes the FormButtons component to the IssueForm component', () => {
        expect(wrapper.find('IssueForm').children().name()).toEqual('FormButtons');
    });

    it('passes the submit button prop to the FormButtons component', () => {
        expect(wrapper.find('FormButtons').prop('submit')).toEqual(true);
    });

    it('passes the clear all button text prop to the FormButtons component', () => {
        expect(wrapper.find('FormButtons').prop('cancelButtonText')).toEqual('Clear All');
    });

    it('passes the clear all button id to the FormButtons component', () => {
        expect(wrapper.find('FormButtons').prop('cancelButtonID')).toEqual('clearAllButton');
    });
});