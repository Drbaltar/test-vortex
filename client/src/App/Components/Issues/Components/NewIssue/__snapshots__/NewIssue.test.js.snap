// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewIssue renders correctly 1`] = `
<SubmissionBody
  submitMapping={
    Object {
      "submitButton": Object {
        "requestFunction": [Function],
        "requestURI": "/api/issues/submit",
      },
    }
  }
>
  <IssueForm
    title="Submit New Issue"
  >
    <FormButtons
      cancelButtonID="clearAllButton"
      cancelButtonText="Clear All"
      submit={true}
    />
  </IssueForm>
</SubmissionBody>
`;
